Add missing include and library paths

carterbox opened the pull request:
Adds the host requirements prefix include-paths and linking-search-paths to the NVCC profile for conda-build (because those two prefixes are separate) by using an activation script only when conda-build is active to set some environment variables which are picked up by the nvcc.profile.

Outside of the conda-build context, these paths are not needed because nvcc would share a prefix with all the cuda libs.

The activation script is located in [cuda-nvcc-feedstock](https://github.com/conda-forge/cuda-nvcc-feedstock/pull/19/)



I didn't do anything for Windows because it's skipped.

<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->
Requires https://github.com/conda-forge/cuda-nvcc-feedstock/pull/19/

carterbox committed:
Add missing include and library paths

jakirkham commented:
cc @robertmaynard

carterbox committed:
REF: Have separate activate script instead of sharing

carterbox committed:
MAI: Bump build number

carterbox commented:
@conda-forge-admin, please rerender

carterbox committed:
BUG: Install activate script

robertmaynard commented on recipe/activate.sh:
```diff
@@ -0,0 +1,25 @@
+if [ "${CONDA_BUILD:-0}" = "1" ]; then
+    CUDA_INCLUDE_DIRS="${CUDA_INCLUDE_DIRS} -I${PREFIX}/${targetsDir}/include"
+    CUDA_INCLUDE_DIRS="${CUDA_INCLUDE_DIRS} -I${BUILD_PREFIX}/${targetsDir}/include"
+    CUDA_LINK_DIRS="${CUDA_LINK_DIRS} -L${PREFIX}/${targetsDir}/lib"
```
Either `PREFIX` or `BUILD_PREFIX` is not needed as that is already added in the nvcc.profile ( `-I$(TOP)/include` ). @jakirkham Do you remember which one `TOP` points too?

carterbox replied:
```
TOP              = $(_HERE_)/../$(_TARGET_DIR_)
```
This would become `$BUILD_PREFIX/$(_TARGET_DIR_)`. I'm not sure whether `$(_TARGET_DIR_)` is equivalent to the `$targetsDir` defined here.

robertmaynard replied:
Yes, `$(_TARGET_DIR_)` will match `targetsDir`

robertmaynard commented on recipe/activate.sh:
```diff
@@ -0,0 +1,25 @@
+    CUDA_LINK_DIRS="${CUDA_LINK_DIRS} -L${BUILD_PREFIX}/${targetsDir}/lib"
+    CUDA_LINK_DIRS="${CUDA_LINK_DIRS} -L${BUILD_PREFIX}/${targetsDir}/lib/stubs"
+else
+    CUDA_INCLUDE_DIRS="${CUDA_INCLUDE_DIRS} -I${CONDA_PREFIX}/${targetsDir}/include"
```
The else isn't need at all due to the existing logic in `nvcc.profile`.

carterbox replied:
You mean that only the IF block is needed. The ELSE block is redundant with `nvcc.profile`?

carterbox committed:
BUG: No deactivation script

carterbox committed:
Rename environment variable to avoid CMake collision

carterbox committed:
DOC: Add a note about why these paths are added in an activation script

carterbox committed:
BUG: Use target_platform instead of cross_target_platform

carterbox commented on recipe/activate.sh:
```diff
@@ -0,0 +1,20 @@
+
+set -ex
+
+[[ "@target_platform@" == "linux-64" ]] && targetsDir="targets/x86_64-linux"
+[[ "@target_platform@" == "linux-ppc64le" ]] && targetsDir="targets/ppc64le-linux"
+[[ "@target_platform@" == "linux-aarch64" ]] && targetsDir="targets/sbsa-linux"
```
I'm observing that targetsDir is an empty string. This probably means that "@target_platform@" is undefined or doesn't equal any of these linux options. What should I be using instead?

carterbox replied:
Probably `$target_platform`

adibbley replied:
Yeah `$target_platform` should be fine here. The `@` syntax is used for substitution in the other feedstock https://github.com/conda-forge/cuda-nvcc-feedstock/blob/main/recipe/build.sh#L10

carterbox replied:
Interesting. What is the utility of using a template like that then replacing it during install?

carterbox committed:
Use target_platform environment variable

jakirkham commented on recipe/activate.sh:
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
Instead of including the activation script here, should we add these changes to `cuda-nvcc`'s activation script (and just keep the `nvcc.profile.patch` change here)?

carterbox replied:
Then the activation script which defines CUDA_LINK_DIRS and the patch which uses CUDA_LINK_DIRS would be in different feedstocks. That seems confusing.

adibbley replied:
Adding this activation script here should be fine, although I think it needs to go to the `cuda-nvcc-impl` package to follow the original intent: https://github.com/conda-forge/cuda-nvcc-feedstock/issues/12#issuecomment-1533562469

carterbox replied:
I put the activation script with cuda-nvcc-tools because that is the package which contains nvcc.profile. nvcc.profile is the only file that needs the activation script, so it would seem odd to separate the two.

carterbox commented on recipe/activate.sh:
```diff
@@ -0,0 +1,23 @@
+
+set -ex
+
+# For conda-build we add the host requirements prefix to the include and link
+# paths because it is separate from the build prefix where nvcc is installed
```
```suggestion
# For conda-build we add the host requirements prefix to the include and link
# paths because it is separate from the build prefix where nvcc is installed
# These environment variables are used in the nvcc.profile patch
```

carterbox committed:
REF: No activation scripts allowed in this feedstock

The complementary part exists in
https://github.com/conda-forge/cuda-nvcc-feedstock/pull/19
